### ISSUE TITLE
Scripts fail to load with a parse error on export projects if it uses editor classes

### ISSUE THREAD

**Issue by ydeltastar** [godotengine/godot#91713]
### Tested versions

v4.3.dev.custom_build [2042420bd]

### System information

Windows 11

### Issue description

Tools scripts using an editor class such as `EditorInterface` will fail to load with a parse error on exported projects.

```gdscript
@tool
extends Node3D

func _init() -> void:
	if Engine.is_editor_hint():
		var editor_filesystem = EditorInterface.get_resource_filesystem()
		if not editor_filesystem.resources_reimported.is_connected(_on_resource_reimport):
			editor_filesystem.resources_reimported.connect(_on_resource_reimport)

func _on_resource_reimport(resources:PackedStringArray):
	pass
```

Output of debug template console wrapper:
```cmd
SCRIPT ERROR: Parse Error: Identifier "EditorInterface" not declared in the current scope.
          at: GDScript::reload (res://node_3d.gd:11)
ERROR: Failed to load script "res://node_3d.gd" with error "Parse error".
   at: ResourceFormatLoaderGDScript::load (modules\gdscript\gdscript.cpp:2894)
```

Well, it makes sense. Editor classes aren't included in distribution builds but there is no conditional compiling in GDScript so tool scripts with game and editor code can't be used at all in a project.

Workarounds:
* Reject types, return to dynamic: `var editor_interface = Engine.get_singleton("EditorInterface")`
* Distribute your source project with `godot.exe`

### Steps to reproduce

Export a project that uses the script above with a debug template and run the `.console.exe` version.

### Minimal reproduction project (MRP)

N/A

**Comment by AThousandShips** [comment]
Related:
* https://github.com/godotengine/godot-proposals/issues/7375
* https://github.com/godotengine/godot-proposals/discussions/4023

**Comment by KoBeWi** [comment]
> Reject types, return to dynamic: var editor_interface = Engine.get_singleton("EditorInterface").

IMO this is the way. You can do this indirectly for partial safety:

```GDScript
func _ready():
	if Engine.is_editor_hint():
		var editor = load("res://some_script_using_editor_interface.gd")
		editor.connect_reimported(_on_resource_reimport)
```
where the loaded script does what you quoted above. It can provide a type-safe interface with only the script in the scene doing unsafe calls.

**Comment by dalexeev** [comment]
> It can provide a type-safe interface with only the script in the scene doing unsafe calls.

This will not work with static typing, only with dynamic typing (if you use `Engine.get_singleton()` and so on for **all** editor classes).

As a workaround, I would suggest my plugin [gdscript-preprocessor](https://github.com/dalexeev/gdscript-preprocessor) (sorry for the self-promotion).

We could try to solve this out of the box by adding an exception for editor classes when analyzing/compiling the script. However, at runtime this should still give an error if execution reaches this point. So I think that this is a rather complex solution, which also adds inconsistency.

**Comment by KoBeWi** [comment]
> This will not work with static typing, only with dynamic typing (if you use Engine.get_singleton() and so on for all editor classes).

No I mean this (for `some_script_using_editor_interface.gd` from my example):
```GDScript
func connect_reimported(callback: Callable):
	var editor_filesystem := EditorInterface.get_resource_filesystem()
	if not editor_filesystem.resources_reimported.is_connected(callback):
		editor_filesystem.resources_reimported.connect(callback)
```
This code is type-safe, but it won't cause problems, because it's loaded using `load()`. The parent script does a dynamic call on this method. Hence I said "partially safe".

**Comment by ydeltastar** [comment]
Also, this isn't mentioned in the documentation and works when running from the editor leaving it to be discovery possible late in development only after exporting.

**Comment by KoBeWi** [comment]
Also related: #56798 (I just remembered)

**Comment by ydeltastar** [comment]
Also: https://github.com/godotengine/godot/issues/86013

**Comment by lucassene** [comment]
Got this happening with me on exported games on 4.3.

The first workaround suggested by @ydeltastar , worked so far.

**Comment by TokisanGames** [comment]
Just discovered this. I have tool scripts in my game that do stuff in the editor, and already filter their actions based on the editor hint. We made EditorInterface static #75694 to make tool scripts easier. But to break an exported game because a script mentions, but doesn't use the class isn't good. 

What about replacing editor classes with dummy classes?

How about recognizing a code block that starts with `if Engine.is_editor_hint():` and not including it in the export?

What about giving us preprocessor defines that demark editor only code, to replace `if Engine.is_editor_hint()` entirely?

**Comment by dalexeev** [comment]
> Python probably replaces it with a placeholder or an error opcode.

Python is a purely dynamic language, type hints are completely optional. Therefore, any non-existent identifiers do not cause a runtime error until execution reaches them.

GDScript combines the qualities of both dynamic and static typing. GDScript uses type hints and implicitly inferred types for static analysis and optimizations, and reports non-existent identifiers (but this does not apply to attribute/index access). Changing this behavior may be undesirable because it helps users find errors faster.

You can use `Engine.get_singleton(&"EditorInterface")` instead of `EditorInterface`, as documented in the description of the new `@export_tool_button` annotation. Similar techniques can be used for other cases besides accessing editor singletons.

But I agree that this is a pretty important point that can hinder users who want to have different code for different platforms, operating systems, distribution stores and other types of engine builds and custom feature tags. I think there are the following options to solve this problem:

1. Introduce a conditional compilation system in the form of preprocessor directives or another mechanism. This is quite complicated, but could potentially be useful for other scenarios.
2. Change the analyzer and compiler behavior only on export and in the exported project. Instead of a compile-time error, generate a "get global" opcode.
3. Replace the error about a non-existent identifier with a warning, which is treated as an error by default (for compatibility). I think this option is most consistent with the principles of GDScript.

**Comment by aekobear** [comment]
This bug is exceedingly hard to uncover with inherited classes. Take this example:

```gdscript
@tool
extends Node
class_name BaseTool

func editor_selection():
  EditorInterface.get_selection()
```

```gdscript
@tool
extends BaseTool
class_name CustomTool

func _ready():
  print("CustomTool is working!");
```

If you create a scene with a single `CustomTool` node, it works perfectly in the editor. But when you export, you get the error:

```
SCRIPT ERROR: Parse Error: Could not resolve class "BaseTool".
          at: GDScript::reload (res://custom-tool.gd:0)
ERROR: Failed to load script "res://custom-tool.gd" with error "Parse error".
   at: load (modules/gdscript/gdscript.cpp:2775)
```

There's no mention of `EditorInterface` at all. Combined with the fact that this functionality only happens on export and is entirely undocumented, basically the only way to uncover the problem is to comment out lines of code one at a time until the problem goes away.

**Comment by ExplodingImplosion** [comment]
> How about recognizing a code block that starts with `if Engine.is_editor_hint():` and not including it in the export?
> 
> What about giving us preprocessor defines that demark editor only code, to replace `if Engine.is_editor_hint()` entirely?

I like this. IMO it would be nice to have a feature that functions similarly to an assert in being fully removed in certain builds. However, I do think that it'd be best to have it as an _alternative_, rather than a replacement, and to provide proper documentation for their differences, so that problems like this are avoided:
> * https://github.com/godotengine/godot-docs/issues/2607

**Comment by gaheldev** [comment]
I got the same issue and found a good workaround, for example:
```gdscript
if Engine.is_editor_hint():
    var undo_redo: UndoRedo = Engine.get_singleton(&"EditorInterface").get_editor_undo_redo()
    shape.owner = Engine.get_singleton(&"EditorInterface").get_edited_scene_root()
```

**Comment by Nodragem** [comment]
I stumble on this today! 😄 In my release build, no dialogue would open in the game because the DialogueOpener class did not compile because of a reference to `EditorInterface`.

What is usually very good with Godot is that once a project works when run from the editor it also works in the release build. 

Because of the `if Engine.is_editor_hint():`, it is clear that the user is not intending to use Editor classes in the release build, so there should be no Error raised, not even a Warning, and most importantly, the game should run as normal (which is not true if a Error is raised).

- Could Godot just remove what is under a `if Engine.is_editor_hint():` block when making the build? 
- If not, could Godot just ignore `EditorInterface` if it is under a `if Engine.is_editor_hint():` block when making the build? 

The solution of raising a warning rather than an error may also be okay in terms of not impacting the game, but really, the user did not do anything wrong. Hope this can be solved and save the next user a few hours 😊

**Comment by FractureFalcon** [comment]
hi, not much to contribute, just wanted to mention that i also had this issue. the solution for this might be a greater discussion on how to exclude code in godot between e.g. debug and release builds. in the meantime, i think a good mitigation may be to include some information on this potential issue and the workaround in the documentation, maybe in the docs for "Running Code in the Editor"? https://docs.godotengine.org/en/stable/tutorials/plugins/running_code_in_the_editor.html#
